sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/ChildPartitionsRecordAction.java

```
@@ -66,63 +66,100 @@ public ChildPartitionsRecordAction(PartitionMetadataDao partitionMetadataDao) {
-              } else {
-                LOG.info(
-                    "["
-                        + token
-                        + "] At least one parent is not finished ("
-                        + "finishedParents = "
-                        + finishedParents
-                        + ", "
-                        + "expectedToBeFinished = "
-                        + (childPartition.getParentTokens().size() - 1)
-                        + "), skipping child partition insertion");
-              }
-
-              return null;
-            });
+        processMerge(partition, record, childPartition);
       }
     }
 
     LOG.info("[" + token + "] Child partitions action completed successfully");
     return Optional.empty();
   }
 
+  private void processSplit(
+      PartitionMetadata partition, ChildPartitionsRecord record, ChildPartition childPartition) {
+    final String partitionToken = partition.getPartitionToken();
+    final String childPartitionToken = childPartition.getToken();
+    LOG.info("[" + partitionToken + "] Processing child partition split event");
+    PARTITION_SPLIT_COUNTER.inc();
+
+    final PartitionMetadata row =
+        toPartitionMetadata(
+            record.getStartTimestamp(),
+            partition.getEndTimestamp(),
+            partition.getHeartbeatMillis(),
+            childPartition);
+    // FIXME: Figure out what to do if this throws an exception
+    partitionMetadataDao.runInTransaction(
+        transaction -> {
+          if (transaction.exists(childPartitionToken)) {
+            LOG.info(
+                "["
+                    + partitionToken
+                    + "] Child partition token "
+                    + childPartitionToken
+                    + " already exists, skipping");
+          } else {
+            LOG.info(
+                "[" + partitionToken + "] Inserting child partition token " + childPartitionToken);
+            transaction.insert(row);
```

What will happen if we insert the same again? Will it give an error due to the same primary key? If we just let it fail and catch the error, would it avoid to call `select count(*)`?

---

We could do that to avoid the query.

---

Changed the implementation to catch the exception instead.